feat(gerrit): implement pr cache

felipecrs opened the pull request:
## Changes

This PR implements [PR cache](https://github.com/renovatebot/renovate/issues/33290) for the Gerrit platform, which delivers massive performance improvements.

PR cache in Gerrit works similarly to other platforms:

- Fetches all Renovate changes and stores them as `PR`s in the repo cache
- The repo cache may be [persisted](https://docs.renovatebot.com/self-hosted-configuration/#repositorycache) between runs
- When there's no cache, the initial sync fetches all Renovate changes through the Gerrit API, 100 changes per page
- When there's cache, the sync fetches changes until it hits a change whose `updated` timestamp matches what's already cached, 20 changes per page. **Gerrit API returns changes ordered by their update date**, meaning no further changes would have been modified.
- The PR cache is only synced once per execution
- When a change is created, updated, or merged, the PR cache for Gerrit is updated on the fly without requiring a re-sync.

This PR is a follow-up to the PR below, which completes the optimization cycle for Gerrit:

- https://github.com/renovatebot/renovate/pull/38966

[Here](https://github.com/renovatebot/renovate/pull/35676#issuecomment-4195881545)'s a comparative benchmark.

## Context

Please select one of the below:

- [ ] This closes an existing Issue:
- [x] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

Did you use AI tools to create any part of this pull request?

Please select one option and, if yes, briefly describe how AI was used (e.g., code, tests, docs) and which tool(s) you used.

- [ ] No — I did not use AI for this contribution.
- [ ] Yes — minimal assistance (e.g., IDE autocomplete, small code completions, grammar fixes).
- [x] Yes — substantive assistance (AI-generated non‑trivial portions of code, tests, or documentation).
- [ ] Yes — other (please describe):

I initially wrote this PR by myself, but then used Claude Opus 4.6 to rewrite it following the PR cache pattern from other platforms. It was also used to implement tests and to help review the code.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

The public repository: https://review.gerrithub.io/q/project:felipecrs/renovate-repro-spring-framework

felipecrs committed:
feat(gerrit): save tons of api requests by caching changes in memory

RahulGautamSingh commented on lib/modules/platform/gerrit/pr-cache.ts:
```diff
@@ -0,0 +1,83 @@
+    }
+    matches.push(pr);
+  }
+  if (limit === 1) {
+    return null;
+  }
```
What does this do?

RahulGautamSingh reviewed (commented):
Thanks for this.

And, as you mentioned this could reuse some logic from the pr-cache implementation in other platforms.

felipecrs committed:
Rework into PrCache style

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-cache

felipecrs committed:
Properly implement reconciliation in PrCache

felipecrs committed:
Restore tests

felipecrs committed:
Fix some issues

felipecrs committed:
Fix existing tests

felipecrs committed:
Change from PrCache to ChangeCache (Gerrit-specific)

felipecrs committed:
perf(gerrit): remove custom scm functions

felipecrs committed:
fix(gerrit): remove memCache to fix prNo coming as null when pr-created

felipecrs committed:
fix(gerrit): remove memCache to fix prNo coming as null when pr-created

felipecrs committed:
Revert "Change from PrCache to ChangeCache (Gerrit-specific)"

This reverts commit 909c3a5dda9a751ba15e53c35b85907481de68a1, as I've
found a cleaner way to achieve these optimizations (reworking Gerrit
SCM).

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-cache

felipecrs committed:
Merge branch 'remove-gerrit-memcache' into gerrit-cache

felipecrs committed:
Remove refreshPr

felipecrs committed:
Merge branch 'remove-gerrit-memcache' into gerrit-cache

felipecrs committed:
Merge branch 'remove-gerrit-memcache' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Fully remove login parameter from configureScm()

felipecrs committed:
Fetch refspecs in batches

felipecrs committed:
Remove unnecessary error handling

felipecrs committed:
Add tests

felipecrs committed:
Throw if branch init fails

felipecrs committed:
Simplify tests a little

felipecrs committed:
Reduce more the tests

felipecrs committed:
More test changes

felipecrs committed:
Reuse the existing test setup more

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-cache

felipecrs committed:
Undo test import change

felipecrs committed:
Improve initializeBranchesFromRefspecs JSDoc

felipecrs committed:
The same for deleteBranchCreatedFromRefspec

felipecrs committed:
Reduce a log to trace level

felipecrs committed:
Merge branch 'simplify-gerrit-scm' into gerrit-cache

felipecrs committed:
Abandon Code-Review -2 changes before initializing repository

felipecrs committed:
Simply initialize the branches after git.initRepo()

felipecrs committed:
Merge branch 'simplify-gerrit-scm' into gerrit-cache

felipecrs committed:
Add note about not reussing GerritPrCache

felipecrs committed:
Rework Pr Cache to be more accurate

felipecrs committed:
Rework approach to be more generic

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' into gerrit-cache

felipecrs committed:
fix(gerrit): avoid retrieving change messages in updatePr and createPr

felipecrs committed:
Remove more of now unnecessary code

felipecrs committed:
Merge branch 'fix-gerrit-pr-body-hash' into gerrit-cache

felipecrs committed:
Add TODO comment about commitAndPush()

felipecrs committed:
fix(gerrit): defer change creation to `createPr()`

Signed-off-by: Felipe Santos <felipe.santos@ericsson.com>

felipecrs committed:
Refactor hashtags handling

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Add prTitle to all missing commitAndPush()

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Fix type check

felipecrs committed:
Fix test issue

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Fix gerrit tests

felipecrs committed:
Add warning when prTitle is set in updatePr

felipecrs committed:
Remove now useless labels from commitAndPush

felipecrs committed:
Fix gerrit client coverage

felipecrs committed:
Fix gerrit scm coverage

felipecrs committed:
Fix rebase coverage

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Fix tests

felipecrs committed:
Improve TODO about assignees

felipecrs committed:
Merge branch 'main' into gerrit-fix-create-pr

felipecrs committed:
Fix formatting

felipecrs committed:
Fix review comment

felipecrs committed:
Merge branch 'main' into gerrit-fix-create-pr

felipecrs committed:
Extract to a pushForReview function

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Fix after #41382

felipecrs committed:
Fix onboarding

felipecrs committed:
Cleanup code

felipecrs committed:
Fix ESLint

felipecrs committed:
Merge branch 'main' into gerrit-fix-create-pr

felipecrs committed:
Fix tests

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Restore gitRetry for git.merge

felipecrs committed:
Fix tests

felipecrs committed:
Simplify hashtags code

felipecrs committed:
Delete pendingChangeBranch in pushForReview()

felipecrs committed:
Merge branch 'main' into gerrit-fix-create-pr

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'main' into gerrit-fix-create-pr

felipecrs committed:
Fix test name

felipecrs committed:
Remove unnecessary test comments

felipecrs committed:
Improve test name again

felipecrs committed:
Fix another test name

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Delete local branch in Gerrit SCM

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Add test for deleting pending branch

felipecrs committed:
Add tests for pushForReview()

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Simplify/optimize commitAndPush

felipecrs committed:
Align jsdoc terminology on virtual branches

felipecrs committed:
Remove no longer relevant TODO

felipecrs committed:
Update virtal branch on push

felipecrs committed:
Simplify deleteVirtualBranch

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Fix test case name

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Improve another test name

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Restore test deleted by mistake

felipecrs committed:
Fix another small test inconsistency

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Fix tests

felipecrs committed:
Extract makeChange test helper to reduce code

felipecrs committed:
Fix coverage

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Fix test coverage

felipecrs committed:
Clean up some tests

felipecrs committed:
Fix test coverage

felipecrs commented:
This PR is in a good, functional state.

But it shouldn't be reviewed as it currently depends on these other two to be merged first:

- https://github.com/renovatebot/renovate/pull/39250
- https://github.com/renovatebot/renovate/pull/38966

felipecrs committed:
Remove `refreshPr()` as we always update cache after modifying change

felipecrs committed:
Remove useless `forceRefresh()`

felipecrs committed:
Merge branch 'main' into gerrit-fix-create-pr

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Simplify Gerrit commitAndPush by leveraging git.commitFiles

felipecrs committed:
Fix Renovate not detecting modified commits

felipecrs committed:
Fix some code style issues

felipecrs committed:
Merge branch 'main' of https://github.com/felipecrs/renovate into gerrit-fix-create-pr

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Revert "Simplify Gerrit commitAndPush by leveraging git.commitFiles"

This reverts commit d71edcc27942327b23003eacdf807c6033592797.

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Fix another small code style issue

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Fix some other code style issues

felipecrs committed:
Use isUndefined instead

felipecrs committed:
Fix one small test style issue

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Fix more code style issues

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Fix yet another small code style issue

felipecrs commented:
Hey @RahulGautamSingh, could you please remove the _auto:inactive-pr_ label from this PR?

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-fix-create-pr

felipecrs committed:
Merge branch 'gerrit-fix-create-pr' of https://github.com/felipecrs/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Fix Gerrit mergeToLocal trying to fetch (and probably failing)

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Fix coverage

felipecrs committed:
Remove handled TODO

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Extract refs/remotes/origin to a function

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Avoid adding prBody message unless it have changed

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Rename virtualBranchRef to remoteBranchRef

felipecrs committed:
Fix remoteBranchRef tsc

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Consistently use currentRevision in Gerrit tests

felipecrs committed:
Remove unnecessary type assertion

felipecrs committed:
Instrument "fetch virtual branches"

felipecrs committed:
Remove pendingChangeBranches, avoid overriding deleteBranch and mergeToLocal

felipecrs committed:
Optimize virtual branches fetch (no longer requires update-ref), convert virtualBranches to Record, rename setVirtualBranch to registerVirtualBranch

felipecrs committed:
Combine registerVirtualBranch and updateVirtualBranch into a single setVirtualBranch, and update ref in virtual branch too

felipecrs committed:
Fix virtualBranches not being initialized

felipecrs committed:
Add some debug logs

felipecrs committed:
Simplify mergeToLocal for virtual branches

felipecrs committed:
Revert "Simplify mergeToLocal for virtual branches"

This reverts commit 375d57a4747f84a149b6a9a436445e43076dcbd6.

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Remove useless isVirtual const

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Merge branch 'main' into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' into gerrit-cache

felipecrs committed:
test(git): add toThrow message for virtualBranches fetch error

Satisfy vitest/require-to-throw-message so oxlint passes.

felipecrs committed:
test(git): add toThrow message for virtualBranches fetch error

Satisfy vitest/require-to-throw-message so oxlint passes.

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

RahulGautamSingh reviewed (commented):
Conflicted.

@felipecrs are you still working on it? If it will be ready in a while, you should draft it.

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into simplify-gerrit-scm

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Update lib/util/git/types.ts

Co-authored-by: RahulGautamSingh <rahultesnik@gmail.com>

felipecrs committed:
Merge branch 'simplify-gerrit-scm' of https://github.com/felipecrs/renovate into gerrit-cache

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-cache

Resolve conflicts from #38966 landing on main (incl. review feedback from
018a06963): keep GerritPrCache helpers/comments on this branch, take
fakeSha/toLongCommitSha test style from main.

felipecrs committed:
fix(gerrit): use static logger messages for cache misses

Satisfy renovate/logger-static-message so WARN logs keep a static msg
for metrics grouping.

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-cache

felipecrs committed:
Fix oxlint

RahulGautamSingh commented on lib/modules/platform/gerrit/index.ts:
```diff
@@ -214,50 +223,97 @@ export async function initRepo({
+    pr.bodyStruct = {
+      hash: newBodyHash,
+    };
+    pr.updatedAt = new Date().toISOString();
```
Two problems here:
- Raw Date violates the Luxon rule in `best-practices.md`
- It produces `…000Z`, but everywhere else the cache stores Gerrit's format via `convertGerritDateToISO (…000000000, no Z)`. reconcile compares `updatedAt.replace('T',' ') === change.updated`, so a body-updated PR never matches on the next sync and its early-stop is skipped. It self-heals (the change just gets re-mapped), so not a correctness bug, but the format diverges. The abandon and merge branches already do the right thing using `change.updated` from the server response. `addMessage` returns void, so there's no server timestamp here, but at minimum use `DateTime` and keep the format consistent.

felipecrs replied:
Thank you. Got rid of that entirely via 3406750d74f154966eccb5813319054eef1542a6.

felipecrs replied:
Actually, I had to restore it. After testing this in production I realized that only Gerrit >=3.10 includes the update date in the response.

https://github.com/renovatebot/renovate/pull/35676/commits/5afb577098a4d6c756b987a09b217acc6f82488b

RahulGautamSingh commented on lib/modules/platform/gerrit/pr-cache.spec.ts:
if a Renovate change exists on the server and isn't in the cache for any reason, a merge silently fails. Is this no-fallback behavior intended, or should these fall back to a direct lookup?

felipecrs replied:
Well, it's not actually silent. It emits a warning.

And that situation should never happen, so I don't think implementing a fallback here is necessary.

RahulGautamSingh commented on lib/modules/platform/gerrit/client.spec.ts:
```diff
@@ -245,24 +245,6 @@ describe('modules/platform/gerrit/client', () => {
-      ).resolves.toEqual([gerritChange({ _number: 1 })]);
-    });
-
     it('sets query.S as 0 if startOffset is not provided', async () => {
```
'sets query.S as 0 if startOffset is not provided' still names the removed startOffset. Rename to "on first page".

felipecrs replied:
Thank you. Done.

viceice commented on lib/modules/platform/gerrit/pr-cache.ts:
```diff
@@ -0,0 +1,165 @@
+    const isSynced = memCache.get<true | undefined>('gerrit-pr-cache-synced');
+
+    if (!isSynced) {
+      await res.sync();
```
this could run twice it init is called in parallel for some reason. use mutex or lazy

felipecrs replied:
@viceice, that is theoretically true, but given Renovate doesn't parallelize any platform call, this doesn't really happen in practice.

Also, this is following the pr-cache logic from all other platforms. All of them are subject to the same issue.

How about you log this as a future improvement instead? (not restricted to Gerrit)

felipecrs committed:
test(gerrit): rename pagination test after startOffset removal

The test still referred to the removed startOffset option.

felipecrs committed:
Return updated change in addMessage

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-cache

felipecrs committed:
Fix addMessage on Gerrit <3.10

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-cache

RahulGautamSingh commented on lib/modules/platform/gerrit/index.ts:
```diff
@@ -305,50 +365,74 @@ export async function createPr(prConfig: CreatePRConfig): Promise<Pr | null> {
+  const cached = prs.find((pr) => pr.number === mergeConfig.id);
+  if (!cached) {
+    logger.warn({ number: mergeConfig.id }, 'mergePr: PR not found in cache');
+    return false;
```
when the PR is missing from cache this returns false without calling submitChange, so a stale cache turns a real merge into a silent failure Renovate will misreport. previously this hit Gerrit directly. is cache always populated?

felipecrs replied:
My initial mindset was to avoid making requests to Gerrit, given our cache **should** be populated correctly.

But in this case we already have the PR/change number, which is a big indicator that the change exists, and if it's not in cache, it's an issue with the cache. It's harmless to self-heal in such case.

So, thanks for the feedback. Implemented via https://github.com/renovatebot/renovate/pull/35676/commits/a2a72a91bb370b3edc8e7cf882d6e94b0724ecf4.

This aligns the implementation with forgejo/gitea as well.

RahulGautamSingh commented on lib/modules/platform/gerrit/pr-cache.ts:
```diff
@@ -0,0 +1,165 @@
+      // If this change exists in cache and has the same updated timestamp,
+      // it hasn't been modified. Since changes are sorted by update time,
+      // all subsequent changes are also unchanged.
+      if (cachedUpdated === change.updated) {
```
this stops mid-page, unlike bitbucket-server which finishes reconciling the already-fetched page and only skips the next page. Gerrit's update-desc sort makes it safe in practice, but a never-cached older change on the same page gets skipped until the next run. consider matching bitbucket server's pattern

felipecrs replied:
Yeah, this is a good idea. The page's data is already there anyway. 83cdf0bb80ec7ecc227a3b5a42e22e335c854351

RahulGautamSingh commented on lib/modules/platform/gerrit/pr-cache.ts:
```diff
@@ -0,0 +1,165 @@
+
+    for (const change of changes) {
+      const cachedPr = items[change._number];
+      const cachedUpdated = cachedPr?.updatedAt?.replace('T', ' ');
```
this hand-reverses convertGerritDateToISO. prefer convertGerritDateToISO(change.updated) === cachedPr.updatedAt so the format logic stays in one place.

felipecrs replied:
That is so much better. Thanks for catching it. 7833987ac23b5a755277287623b00cefcae9a35f

RahulGautamSingh commented on lib/modules/platform/gerrit/pr-cache.spec.ts:
```diff
@@ -0,0 +1,259 @@
+
+      const pr1 = mapGerritChangeToPr(makeChange({ _number: 1 }))!;
+      const pr2 = mapGerritChangeToPr(makeChange({ _number: 2 }))!;
+      pr1.updatedAt = undefined as never;
```
`as never` forces undefined onto a required field. use partial<Pr>() instead.

felipecrs replied:
Turns out I don't need `as never` since updatedAt is already optional in `Pr`.

felipecrs committed:
Reconcile entire page rather than stopping at first match

felipecrs committed:
Remove `as never` from test

felipecrs committed:
Reuse convertGerritDateToISO instead

felipecrs committed:
Attempt to get change from Gerrit when a number is provided

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-cache

RahulGautamSingh commented on lib/modules/platform/gerrit/index.ts:
```diff
@@ -305,50 +390,68 @@ export async function createPr(prConfig: CreatePRConfig): Promise<Pr | null> {
     sourceBranch: prConfig.sourceBranch,
     prBody: prConfig.prBody,
-  });
+  })!;
```
the `!` hides that `mapGerritChangeToPr` can return `null` when `sourceBranch` is missing. check the result and throw a clear error instead of asserting non-null.

felipecrs replied:
It's impossible that `sourceBranch` is missing since we pass it as parameter. Checking and throwing the error here would just make the code more complex.

Please let me know if you'd like me to implement it anyway.

RahulGautamSingh replied:
Yes, I got that it will be impossible since you explicitly pass it. Its okay to ignore this.

RahulGautamSingh commented on lib/modules/platform/gerrit/index.ts:
```diff
@@ -214,50 +223,126 @@ export async function initRepo({
+      !isUndefined(findPRConfig.state) &&
+      findPRConfig.state !== 'all' &&
+      pr.state !== findPRConfig.state &&
+      !(findPRConfig.state === '!open' && pr.state !== 'open')
```
this state-matching condition is hard to read, especially the `!open` double negative. extract a `matchesState(pr.state, findPRConfig.state)` helper with early returns.

felipecrs replied:
Done via https://github.com/renovatebot/renovate/pull/35676/commits/4d8a9a5646ebfc1a1e2654bbe5620620f993cadb.

felipecrs committed:
Extract prStateMatchesFilter helper

felipecrs committed:
Merge branch 'main' of https://github.com/renovatebot/renovate into gerrit-cache
